test/benchmark/benchmark_malloc_realloc_mimick.cpp

```diff
@@ -0,0 +1,202 @@
+#include "performance_test_fixture/mimick_performance_test_fixture.hpp"
+
+namespace
+{
+
+constexpr int kForcePerformanceTracking = 2;
+constexpr int kEnablePerformanceTracking = 1;
+constexpr int kDisablePerformanceTracking = 0;
+
+}  // namespace
+
+class PerformanceTestFixture : public performance_test_fixture::MimickPerformanceTest
+{
+public:
+  void SetUp(benchmark::State & state)
+  {
+    switch (state.range(0)) {
+      case kDisablePerformanceTracking:
+        break;
+      case kEnablePerformanceTracking:
+        performance_test_fixture::MimickPerformanceTest::SetUp(state);
+        break;
+      case kForcePerformanceTracking:
+        performance_test_fixture::MimickPerformanceTest::SetUp(state);
+        memory_manager->Start();
+        break;
+      default:
+        std::string message =
+          std::string("Invalid enable/disable value: ") + std::to_string(state.range(0));
+        state.SkipWithError(message.c_str());
+        break;
+    }
+  }
+
+  void TearDown(benchmark::State & state)
+  {
+    if (kEnablePerformanceTracking == state.range(0)) {
+      performance_test_fixture::MimickPerformanceTest::TearDown(state);
+    } else if (kForcePerformanceTracking == state.range(0)) {
+      memory_manager->Pause();
+    }
+  }
+};
+
+// This does not make use of PauseTiming or ResumeTiming because timing is very short for these
+// benchmarks. However, they should be enough to measure the overhead of turning on performance
+// measurements.
+BENCHMARK_DEFINE_F(PerformanceTestFixture, benchmark_on_malloc)(
+  benchmark::State & state)
+{
+  const size_t alloc_size = state.range(1);
+  if (alloc_size < 1) {
+    state.SkipWithError("Size for allocation is too small for this test");
+  }
+  for (auto _ : state) {
+    void * ptr = std::malloc(alloc_size);
+    if (PERFORMANCE_TEST_FIXTURE_UNLIKELY(nullptr == ptr)) {
+      state.SkipWithError("Malloc failed to malloc");
+    }
+    std::free(ptr);
+    benchmark::DoNotOptimize(ptr);
+    benchmark::ClobberMemory();
+  }
+}
+
+BENCHMARK_DEFINE_F(PerformanceTestFixture, benchmark_on_calloc)(
+  benchmark::State & state)
+{
+  const size_t alloc_size = state.range(1);
+  if (alloc_size < 1) {
+    state.SkipWithError("Size for allocation is too small for this test");
+  }
+  for (auto _ : state) {
+    void * ptr = std::calloc(1, alloc_size);
+    if (PERFORMANCE_TEST_FIXTURE_UNLIKELY(nullptr == ptr)) {
+      state.SkipWithError("Calloc failed to calloc");
+    }
+    std::free(ptr);
+    benchmark::DoNotOptimize(ptr);
+    benchmark::ClobberMemory();
+  }
+}
+
+BENCHMARK_DEFINE_F(PerformanceTestFixture, benchmark_on_malloc_many)(
+  benchmark::State & state)
+{
+  const size_t alloc_size = state.range(1);
+  if (alloc_size < 1) {
+    state.SkipWithError("Size for allocation is too small for this test");
+  }
+  void * ptrs[4096];
+  for (auto _ : state) {
+    for (size_t i = 0; i < 4096; i++) {
+      ptrs[i] = std::malloc(alloc_size);
+      if (nullptr == ptrs[i]) {
```

Comment:
@cottsay no branch hint? Or does this manage to exhaust memory?

Reply:
Good catch, 5e026e0.